domain/upgrade/state/state.go

```diff
@@ -0,0 +1,304 @@
+		StateBase: domain.NewStateBase(factory),
+	}
+}
+
+func (st *State) getCurrentUpgrade(ctx context.Context, tx *sqlair.TX) (info, []infoControllerNode, error) {
+	q1 := `
+SELECT   (uuid, previous_version, target_version, created_at, started_at) AS &info.* FROM upgrade_info AS info
+WHERE    info.completed_at IS NULL
+ORDER BY info.created_at DESC LIMIT 1`
+	s1, err := sqlair.Prepare(q1, info{})
+	if err != nil {
+		return info{}, nil, errors.Annotatef(err, "preparing %q", q1)
+	}
+
+	q2 := `
+SELECT (controller_node_id, node_upgrade_completed_at) AS &infoControllerNode.*
+FROM   upgrade_info_controller_node
+WHERE  upgrade_info_controller_node.upgrade_info_uuid = $M.info_uuid `
+	s2, err := sqlair.Prepare(q2, infoControllerNode{}, sqlair.M{})
+	if err != nil {
+		return info{}, nil, errors.Annotatef(err, "preparing %q", q2)
+	}
+
+	var upgradeInfo info
+	err = tx.Query(ctx, s1).Get(&upgradeInfo)
+	if err == sqlair.ErrNoRows {
+		return info{}, nil, noUpgradeErr
+	}
+	if err != nil {
+		return info{}, nil, errors.Trace(err)
+	}
+
+	var nodeInfos []infoControllerNode
+	err = tx.Query(ctx, s2, sqlair.M{"info_uuid": upgradeInfo.UUID}).GetAll(&nodeInfos)
+	if err != nil && err != sqlair.ErrNoRows {
+		return info{}, nil, errors.Trace(err)
+	}
+	return upgradeInfo, nodeInfos, nil
+}
+
+// CreateUpgrade creates an upgrade to and from specified versions.
+// If an upgrade is already running/pending, return an error
+func (st *State) CreateUpgrade(ctx context.Context, previousVersion, targetVersion version.Number) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	return errors.Trace(db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		info, _, err := st.getCurrentUpgrade(ctx, tx)
+		if err != nil && err != noUpgradeErr {
+			return errors.Trace(err)
+		}
+		if err == nil {
+			return errors.AlreadyExistsf("upgrade %s -> %s", info.PreviousVersion, info.TargetVersion)
+		}
+
+		q := `
+INSERT INTO upgrade_info (uuid, previous_version, target_version, created_at) VALUES
+($M.uuid, $M.previousVersion, $M.targetVersion, DATETIME('now'))`
+		s, err := sqlair.Prepare(q, sqlair.M{})
+		if err != nil {
+			return errors.Annotatef(err, "preparing %q", q)
+		}
+		return errors.Trace(tx.Query(ctx, s, sqlair.M{
+			"uuid":            utils.MustNewUUID().String(),
+			"previousVersion": previousVersion.String(),
+			"targetVersion":   targetVersion.String(),
+		}).Run())
+	}))
+}
+
+// SetControllerReady marks the supplied controllerID as being ready
+// to start it's upgrade. All provisioned controllers need to be ready
+// before an upgrade can start
+// a controller is ready for a particualry upgrade if it's ID is present
+func (st *State) SetControllerReady(ctx context.Context, controllerID string) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	return errors.Trace(db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		info, nodeInfos, err := st.getCurrentUpgrade(ctx, tx)
```

Comment:
This is a little dicey. I think you should assume that individual controller machines have the UUID of the upgrade they are dealing with, and make it an argument of this method.